File: clearspend/frontend/src/api/transactions.ts
/** Load transaction data for the authenticated dashboard. */

const API_URL = import.meta.env.VITE_API_URL ?? 'http://127.0.0.1:8000'

export type MonthlySummary = {
  month: string
  transaction_count: number
  overall_total: string
  totals_by_category: Record<string, string>
}

export type Transaction = {
  id: number
  amount: string
  merchant: string
  category: string
  date: string
}

export type TransactionPage = {
  items: Transaction[]
  total: number
  page: number
  page_size: number
}

export type TransactionFilters = {
  search: string
  category: string
  start: string
  end: string
  sortBy: 'date' | 'amount' | 'merchant' | 'category'
  sortDirection: 'asc' | 'desc'
}

type FastApiError = {
  detail?: string | Array<{ msg: string }>
}

function errorMessage(error: FastApiError): string {
  if (typeof error.detail === 'string') return error.detail
  if (Array.isArray(error.detail) && error.detail.length > 0) {
    return error.detail[0].msg
  }
  return 'The transaction could not be saved.'
}

export async function getMonthlySummary(
  token: string,
  month: string,
  signal?: AbortSignal,
): Promise<MonthlySummary> {
  const query = new URLSearchParams({ month })
  const response = await fetch(`${API_URL}/transactions/summary?${query}`, {
    headers: { Authorization: `Bearer ${token}` },
    signal,
  })

  if (!response.ok) {
    throw new Error('The monthly summary could not be loaded.')
  }

  return (await response.json()) as MonthlySummary
}

export async function createTransaction(
  token: string,
  amount: string,
  merchant: string,
  date: string,
): Promise<Transaction> {
  const response = await fetch(`${API_URL}/transactions/`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ amount, merchant, date }),
  })

  if (!response.ok) {
    const error = (await response.json()) as FastApiError
    throw new Error(errorMessage(error))
  }

  return (await response.json()) as Transaction
}

export async function updateTransaction(
  token: string,
  transactionId: number,
  changes: { amount: string; merchant: string; date: string },
): Promise<Transaction> {
  const response = await fetch(`${API_URL}/transactions/${transactionId}`, {
    method: 'PATCH',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(changes),
  })

  if (!response.ok) {
    const error = (await response.json()) as FastApiError
    throw new Error(errorMessage(error))
  }

  return (await response.json()) as Transaction
}

export async function deleteTransaction(
  token: string,
  transactionId: number,
): Promise<void> {
  const response = await fetch(`${API_URL}/transactions/${transactionId}`, {
    method: 'DELETE',
    headers: { Authorization: `Bearer ${token}` },
  })

  if (!response.ok) {
    const error = (await response.json()) as FastApiError
    throw new Error(errorMessage(error))
  }
}

export async function getTransactions(
  token: string,
  page: number,
  filters: TransactionFilters,
  signal?: AbortSignal,
): Promise<TransactionPage> {
  const query = new URLSearchParams({
    page: String(page),
    page_size: '10',
    sort_by: filters.sortBy,
    sort_direction: filters.sortDirection,
  })
  if (filters.search) query.set('search', filters.search)
  if (filters.category) query.set('category', filters.category)
  if (filters.start) query.set('start', filters.start)
  if (filters.end) query.set('end', filters.end)
  const response = await fetch(`${API_URL}/transactions/?${query}`, {
    headers: { Authorization: `Bearer ${token}` },
    signal,
  })

  if (!response.ok) {
    throw new Error('Transactions could not be loaded.')
  }

  return (await response.json()) as TransactionPage
}

File: clearspend/app/api/routes/transactions.py
"""Handle transaction creation, browsing, reports, updates, and deletion."""

from datetime import date
from decimal import Decimal
import csv
import io

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from fastapi.responses import StreamingResponse
from sqlalchemy import func, or_
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user, get_db
from app.models.category import Category
from app.models.category_rule import CategoryRule
from app.models.transaction import Transaction
from app.models.user import User
from app.schemas.transaction import (
    MonthlySummary,
    SortDirection,
    TransactionCreate,
    TransactionPage,
    TransactionRead,
    TransactionSortField,
    TransactionUpdate,
)
from app.services.categorizer import categorize_transaction

# Router object is created for transaction related endpoints
router = APIRouter(prefix="/transactions", tags=["transactions"])


def find_category(name: str, owner_id: int, db: Session) -> Category:
    category = (
        db.query(Category)
        .filter(
            Category.name == name,
            or_(Category.is_default.is_(True), Category.owner_id == owner_id),
        )
        .one_or_none()
    )
    if category is None:
        raise HTTPException(status_code=422, detail="category does not exist")
    return category


def categorize_merchant(merchant: str, owner_id: int, db: Session) -> str:
    stored_rules = (
        db.query(CategoryRule)
        .join(CategoryRule.category)
        .filter(
            CategoryRule.is_active.is_(True),
            or_(
                CategoryRule.is_default.is_(True),
                CategoryRule.owner_id == owner_id,
            ),
        )
        .order_by(CategoryRule.priority.asc(), CategoryRule.id.asc())
        .all()
    )
    rules = ((rule.keyword, rule.category.name) for rule in stored_rules)
    return categorize_transaction(merchant, rules)


# Transaction creation endpoint is defined
@router.post("/", response_model=TransactionRead)
def create_transaction(
    transaction: TransactionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    # Category is determined using business logic
    category_name = categorize_merchant(transaction.merchant, current_user.id, db)
    category = find_category(category_name, current_user.id, db)

    # Transaction database object is created
    db_transaction = Transaction(
        amount=transaction.amount,
        merchant=transaction.merchant,
        category_record=category,
        owner=current_user,
        date=transaction.date,
    )

    # Transaction is added to the database session
    db.add(db_transaction)

    # Changes are committed to the database
    db.commit()

    # Database generated values are refreshed
    db.refresh(db_transaction)

    # Stored transaction is returned
    return db_transaction


# Transaction list endpoint is defined
@router.get("/", response_model=TransactionPage)
def list_transactions(
    start: date | None = None,
    end: date | None = None,
    category: str | None = Query(default=None, min_length=1, max_length=100),
    search: str | None = Query(default=None, min_length=1, max_length=200),
    sort_by: TransactionSortField = TransactionSortField.date,
    sort_direction: SortDirection = SortDirection.desc,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=25, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if start is not None and end is not None and start > end:
        raise HTTPException(
            status_code=422,
            detail="start date must be on or before end date",
        )

    # Base query for transactions is created
    query = (
        db.query(Transaction)
        .join(Transaction.category_record)
        .filter(Transaction.owner_id == current_user.id)
    )

    # Start date filter is applied if provided
    if start is not None:
        query = query.filter(Transaction.date >= start)

    # End date filter is applied if provided
    if end is not None:
        query = query.filter(Transaction.date <= end)

    if category is not None:
        query = query.filter(Category.name == category.strip())

    if search is not None:
        query = query.filter(Transaction.merchant.ilike(f"%{search.strip()}%"))

    total = query.count()
    sort_column = getattr(Transaction, sort_by.value)
    order_expression = (
        sort_column.asc()
        if sort_direction == SortDirection.asc
        else sort_column.desc()
    )
    items = (
        query.order_by(order_expression, Transaction.id.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    return {
        "items": items,
        "total": total,
        "page": page,
        "page_size": page_size,
    }


# Monthly summary endpoint is defined
@router.get("/summary", response_model=MonthlySummary)
def monthly_summary(
    month: str = Query(pattern=r"^\d{4}-(0[1-9]|1[0-2])$"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    month_key = month
    year, month_number = (int(part) for part in month_key.split("-"))
    month_start = date(year, month_number, 1)
    next_month = (
        date(year + 1, 1, 1)
        if month_number == 12
        else date(year, month_number + 1, 1)
    )

    # Grouped totals are computed at the database level
    grouped_rows = (
        db.query(
            Category.name.label("category"),
            func.count(Transaction.id).label("transaction_count"),
            func.sum(Transaction.amount).label("category_total"),
        )
        .join(Transaction.category_record)
        .filter(
            Transaction.owner_id == current_user.id,
            Transaction.date >= month_start,
            Transaction.date < next_month,
        )
        .group_by(Category.name)
        .all()
    )

    # Totals by category container is created
    totals_by_category: dict[str, Decimal] = {}

    # Overall total accumulator is created
    overall_total = Decimal("0.00")

    # Transaction count accumulator is created
    transaction_count = 0

    # Grouped rows are converted into the response structures
    for row in grouped_rows:
        # Category label is read
        category = row.category

        category_total = Decimal(row.category_total or 0).quantize(Decimal("0.01"))

        # Transaction count value is normalized
        category_count = int(row.transaction_count or 0)

        # Category total is stored
        totals_by_category[category] = category_total

        # Overall total is accumulated
        overall_total += category_total

        # Overall transaction count is accumulated
        transaction_count += category_count

    # Summary response is returned
    return {
        "month": month_key,
        "transaction_count": transaction_count,
        "overall_total": overall_total,
        "totals_by_category": totals_by_category,
    }


@router.get("/export.csv")
def export_transactions(
    start: date | None = None,
    end: date | None = None,
    category: str | None = None,
    search: str | None = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if start is not None and end is not None and start > end:
        raise HTTPException(status_code=422, detail="start date must be on or before end date")
    query = db.query(Transaction).join(Transaction.category_record).filter(Transaction.owner_id == current_user.id)
    if start is not None: query = query.filter(Transaction.date >= start)
    if end is not None: query = query.filter(Transaction.date <= end)
    if category: query = query.filter(Category.name == category.strip())
    if search: query = query.filter(Transaction.merchant.ilike(f"%{search.strip()}%"))
    output = io.StringIO(); writer = csv.writer(output, lineterminator="\n"); writer.writerow(["date", "merchant", "category", "amount"])
    for transaction in query.order_by(Transaction.date.desc(), Transaction.id.desc()).yield_per(1000):
        writer.writerow([transaction.date.isoformat(), transaction.merchant, transaction.category, f"{Decimal(transaction.amount):.2f}"])
    return StreamingResponse(iter([output.getvalue()]), media_type="text/csv", headers={"Content-Disposition": "attachment; filename=clearspend-transactions.csv"})


def find_transaction(
    transaction_id: int,
    owner_id: int,
    db: Session,
) -> Transaction:
    transaction = (
        db.query(Transaction)
        .filter(
            Transaction.id == transaction_id,
            Transaction.owner_id == owner_id,
        )
        .one_or_none()
    )
    if transaction is None:
        raise HTTPException(status_code=404, detail="transaction not found")
    return transaction


@router.get("/{transaction_id}", response_model=TransactionRead)
def get_transaction(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return find_transaction(transaction_id, current_user.id, db)


@router.patch("/{transaction_id}", response_model=TransactionRead)
def update_transaction(
    transaction_id: int,
    changes: TransactionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    transaction = find_transaction(transaction_id, current_user.id, db)
    update_data = changes.model_dump(exclude_unset=True)
    requested_category = update_data.pop("category", None)

    for field, value in update_data.items():
        setattr(transaction, field, value)

    if requested_category is not None:
        transaction.category_record = find_category(
            requested_category,
            current_user.id,
            db,
        )
    elif "merchant" in update_data:
        category_name = categorize_merchant(
            transaction.merchant,
            current_user.id,
            db,
        )
        transaction.category_record = find_category(
            category_name,
            current_user.id,
            db,
        )

    db.commit()
    db.refresh(transaction)
    return transaction


@router.delete("/{transaction_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_transaction(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Response:
    transaction = find_transaction(transaction_id, current_user.id, db)
    db.delete(transaction)
    db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)
